src/main/java/org/truffleruby/language/loader/FeatureLoader.java

```diff
@@ -238,47 +238,60 @@ public void ensureCExtImplementationLoaded(String feature, RequireNode requireNo
                 throw new RaiseException(context.getCoreExceptions().loadError("Sulong is required to support C extensions, and it doesn't appear to be available", feature, null));
             }
 
-            final String cextRBpath = context.getRubyHome() + "/lib/truffle/truffle/cext.rb";
-            requireNode.executeRequire(cextRBpath);
+            requireNode.executeRequire(context.getRubyHome() + "/lib/truffle/truffle/cext.rb");
 
-            String rubySUpath = loadCExtLibRuby(feature);
+            final String rubySUpath = context.getRubyHome() + "/lib/cext/ruby.su";
+            final List<TruffleObject> libraries = loadCExtLibRuby(rubySUpath, feature);
 
-            sulongLoadLibraryFunction = getCExtFunction("rb_tr_load_library", rubySUpath);
+            sulongLoadLibraryFunction = RequireNode.create().findFunctionInLibraries(libraries, "rb_tr_load_library", rubySUpath);
 
             cextImplementationLoaded = true;
         }
     }
 
-    @TruffleBoundary
-    private String loadCExtLibRuby(String feature) {
-        String rubySUpath = context.getRubyHome() + "/lib/cext/ruby.su";
-
+    private List<TruffleObject> loadCExtLibRuby(String rubySUpath, String feature) {
         if (context.getOptions().CEXTS_LOG_LOAD) {
             Log.LOGGER.info(() -> String.format("loading cext implementation %s", rubySUpath));
         }
 
         if (!new File(rubySUpath).exists()) {
-            throw new RaiseException(context.getCoreExceptions().loadError("This TruffleRuby distribution does not have the C extension implementation file ruby.su", feature, null));
+            throw new RaiseException(context.getCoreExceptions().loadError("this TruffleRuby distribution does not have the C extension implementation file ruby.su", feature, null));
         }
 
-        loadCExtLibrary(rubySUpath);
-
-        return rubySUpath;
+        return loadCExtLibrary(rubySUpath);
     }
 
     @TruffleBoundary
-    public void loadCExtLibrary(String path) {
+    public List<TruffleObject> loadCExtLibrary(String path) {
         File file = new File(path);
 
         if (!new File(path).exists()) {
             throw new RaiseException(context.getCoreExceptions().loadError(path + " does not exists", path, null));
         }
 
+        final List<TruffleObject> libraries = new ArrayList<>();
+
         try {
-            Linker.loadLibrary(file, this::loadNativeLibrary, this::parseSource);
+            Linker.loadLibrary(file, this::loadNativeLibrary, source -> {
+                final Object result;
+
+                try {
+                    result = context.getEnv().parse(source).call();
+                } catch (Exception e) {
+                    throw new JavaException(e);
+                }
+
+                if (!(result instanceof TruffleObject)) {
+                    throw new RaiseException(context.getCoreExceptions().loadError(file + " returned something that isn't an object", path, null));
```

Comment:
This exception message looks like it could be confusing. If I got it I wonder what I received that wasn't an object. I think the old message qualifying it as a Truffle object was more helpful.

Reply:
Fixed.